core/domain/exp_services.py

```diff
@@ -1723,6 +1723,496 @@ def is_voiceover_change_list(change_list):
+            % (frontend_version, backend_version))
+
+    snapshots_metadata = get_exploration_snapshots_metadata(exp_id)
+    composite_change_list_dict = []
+    for i in python_utils.RANGE(frontend_version, backend_version):
+        composite_change_list_dict += snapshots_metadata[i]['commit_cmds']
+
+    composite_change_list = [
+        exp_domain.ExplorationChange(change)
+        for change in composite_change_list_dict]
+
+    return composite_change_list
+
+
+def are_changes_mergeable(exp_id, frontend_version, change_list):
+    """Checks whether the change list can be merged when the
+    frontend version is not equal to the latest backend version.
+
+    Args:
+        exp_id: str. The id of the exploration where the change_list is to
+            be applied.
+        frontend_version: int. Version of an exploration from frontend on
+            which a user is working.
+        change_list: list(ExplorationChange). List of the changes made by the
+            user on the frontend, which needs to be checked for mergeability.
+
+    Returns:
+        boolean. Whether the changes are mergeable.
+    """
+
+    backend_version = exp_fetchers.get_exploration_by_id(exp_id).version
+    if backend_version == frontend_version:
+        return True
+    else:
+        try:
+            # A complete list of changes from one version to another
+            # is composite_change_list.
+            composite_change_list = get_composite_change_list(
+                exp_id, frontend_version, backend_version)
+
+            # Added_state_names: list(str). Names of the states added to the
+            # exploration from prev_exp_version to current_exp_version. It
+            # stores the latest name of the added state.
+            added_state_names = []
+
+            # Deleted_state_names: list(str). Names of the states deleted from
+            # the exploration from prev_exp_version to current_exp_version.
+            # It stores the initial name of the deleted state from
+            # pre_exp_version.
+            deleted_state_names = []
+
+            # New_to_old_state_names: dict. Dictionary mapping state names of
+            # current_exp_version to the state names of prev_exp_version.
+            # It doesn't include the name changes of added/deleted states.
+            new_to_old_state_names = {}
+
+            # Changed_properties: dict. List of all the properties changed
+            # according to the state and property name.
+            changed_properties = {}
+
+            # Changed_translations: dict. List of all the translations changed
+            # according to the state and content_id name.
+            changed_translations = {}
+            frontend_version_exploration = exp_fetchers.get_exploration_by_id(
+                exp_id, version=frontend_version)
+            backend_version_exploration = exp_fetchers.get_exploration_by_id(
+                exp_id, version=backend_version)
```

Comment:
This is already fetched at the top, right? Can't we use the same object?

Reply:
Done.